src/tests/Microsoft.PowerFx.Interpreter.Tests/ServiceProviderTests.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using Xunit;
+
+namespace Microsoft.PowerFx.Interpreter.Tests
+{
+    public class ServiceProviderTests
+    {
+        [Fact]
+        public void Test()
+        {
+            var r1 = new BasicServiceProvider();
+            Assert.Throws<ArgumentNullException>(() => r1.AddService<MyService>(null));
+
+            Assert.Throws<ArgumentNullException>(() => r1.AddService(null, new MyService()));
+
+            // Nulls are ignored, don't crash
+            var r2 = new BasicServiceProvider(null);
+            Assert.Null(r2.GetService(typeof(MyService)));
+
+            r2 = new BasicServiceProvider((IServiceProvider[])null);
+            Assert.Null(r2.GetService(typeof(MyService)));
+
+            r2 = new BasicServiceProvider(new IServiceProvider[0]);
+            Assert.Null(r2.GetService(typeof(MyService)));
+        }
+
+        [Fact]
+        public void ServiceChaining()
+        {
+            var service1 = new MyService();
+            var r1 = new BasicServiceProvider();
+            var otherService = new OtherService();
+
+            r1.AddService(service1);
+            r1.AddService(otherService);
+
+            // Lookup succeeds
+            var lookup = r1.GetService(typeof(MyService));
+            Assert.Same(lookup, service1);
+
+            var r2 = new BasicServiceProvider();
+            var r21 = new BasicServiceProvider(r2, null, r1);
+
+            // Finds in child
+            lookup = r21.GetService(typeof(MyService));
+            Assert.Same(lookup, service1);
+
+            // Shadowing 
+            var service2 = new MyService();
+            Assert.NotSame(service1, service2);
+
+            r2.AddService(service2);
+
+            var lookup1 = r1.GetService(typeof(MyService));
+            Assert.Same(lookup1, service1);
```

Comment:
Is this same as line 41-42?

Reply:
yes - but it's making sure the call on 55 doesn't interfere and we get same results. .